06_gpu_and_ml/reinforcement-learning/trl-grpo.py

```diff
@@ -0,0 +1,169 @@
+        start_idx = completion.find("```python") + len("```python")
+        end_idx = completion.find("```", start_idx)
+        if end_idx != -1:
+            code = completion[start_idx:end_idx].strip()
+        else:
+            code = completion[start_idx:].strip()
+    else:
+        code = completion.strip()
+
+    test_cases = "\n".join(testcase)
+    full_code = f"{code}\n\n{test_cases}"
+    return full_code
+
+
+# Finally, we define the function that is passed into the GRPOTrainer, which takes in a list of completions
+# Custom reward functions must conform to a [specific signature](https://huggingface.co/docs/trl/main/en/grpo_trainer#using-a-custom-reward-function)
+def reward_helper_function(prompts, completions, testcases, **kwargs):
+    return compute_reward.map(zip(completions, testcases))
+
+
+# ## Kicking off a training run
+
+# We import the necessary libraries
+with image.imports():
+    from datasets import load_dataset
+    from trl import GRPOConfig, GRPOTrainer
+
+
+# Preprocess the data, renaming columns to
+def load_and_preprocess_data():
+    dataset = load_dataset(
+        "OpenCoder-LLM/opc-sft-stage2", "educational_instruct", split="train"
+    )
+    dataset = dataset.rename_column(
+        "instruction", "prompt"
+    )  # needed for the GRPO trainer
+    dataset = dataset.rename_column("testcase", "testcases")
+    return dataset
+
+
+# We use WANDB for logging, hence we use a [Modal Secret](https://modal.com/docs/guide/secrets#secrets) with wandb credentials
+@app.function(
+    image=image,
+    gpu="H100",
+    timeout=86400,
+    secrets=[modal.Secret.from_name("wandb-secret")],
+)
+def train():
+    dataset = load_and_preprocess_data()
+    training_args = GRPOConfig(
+        output_dir="Qwen2-0.5B-GRPO", report_to="wandb"
+    )  # comment if don't want to use wandb
+
+    trainer = GRPOTrainer(
+        model="Qwen/Qwen2-0.5B-Instruct",
+        reward_funcs=reward_helper_function,
+        args=training_args,
+        train_dataset=dataset,
+    )
+    trainer.train()
+
+
+# To run: `modal run --detach trl-grpo.py::train``
+
+# ## Speeding up training with vLLM
+
+# vLLM can be used either in server mode (run vLLM server on separate gpu) or colocate mode (within the training process)
+
+@app.function(
+    image=image,
+    gpu="H100:2",
+    timeout=86400,
+    secrets=[modal.Secret.from_name("wandb-secret")],
+)
+def train_vllm_server_mode():
+    dataset = load_and_preprocess_data()
+    subprocess.run(
+        ["trl", "vllm-serve", "--model", "Qwen/Qwen2-0.5B-Instruct"],
+        env={"CUDA_VISIBLE_DEVICES": "0"}, # Run on separate GPU
+    )
+    training_args = GRPOConfig(output_dir="Qwen2-0.5B-GRPO")
+    trainer = GRPOTrainer(
+        model="Qwen/Qwen2-0.5B-Instruct",
+        reward_funcs=reward_helper_function,
+        args=training_args,
+        train_dataset=dataset,
+        use_vllm=True,
+        vllm_mode="server",
+    )
+    trainer.train()
```

Comment:
Should we also limit the visible cuda devices for the trainer process? The default `trl vllm-serve` will have `gpu_memory_utilization=0.9` I think. If that's the case, there won't be much space on cuda device 0 for the trainer, unless I'm misunderstanding and the trainer is only running on cuda device 1 by default. Alternatively, if the trainer only runs on cuda device 0 by default, you could just move the vllm server to cuda device 1.

In either case, I'd lean toward being explicit about this, both in the code and in the commentary.